jpa/org.eclipse.persistence.jpa/src/org/eclipse/persistence/internal/jpa/metadata/accessors/objects/MetadataAsmFactory.java

```diff
@@ -372,14 +376,28 @@ public AnnotationVisitor visitArray(String name) {
 
         @Override
         public void visitEnd() {
-            if (this.element != null) {
-                if (this.annotation.isMeta()) {
-                    this.element.addMetaAnnotation(this.annotation);
+            if ( this.element != null && !isRecursive()  ) {
+                if ( this.annotation.isMeta() ) {
+                    this.element.addMetaAnnotation( this.annotation );
                 } else {
-                    this.element.addAnnotation(this.annotation);
+                    this.element.addAnnotation( this.annotation );
                 }
             }
         }
+        private boolean isRecursive() {
+            if(this.element == null || this.annotation == null)
+                return false;
+            if(this.annotation.getName().equals( this.element.getName())) {
+                //simple recursion (annotation annotated with itself)
+                return true;
+            }
+            MetadataClass aClass = this.element.m_factory.m_metadataClasses.get( this.annotation.m_name );
+            if ( aClass != null && !aClass.m_metaAnnotations.isEmpty() ) {
```

Comment:
I'm not sure if isEmpty() check makes it much faster. It can save hash(key) computing for empty map, but on the other hand it's being called always.